### PR TITLE
os: implement and test os.MkdirTemp

### PR DESCRIPTION
**Pull request by dkegel-fastly**
Also:
- remove one TODO now that CreateTemp exists
- add a testcase for os.stat to expose a problem found by upstream's test for MkdirTemp
- add a forgotten handleSyscallError in os.stat in unix to fix that problem there
- skip tests that expose same problem on Windows for now; it's kind of subtle

### CODE REVIEW

File: src/os/tempfile.go [hunk 1]
```diff
@@ -95,6 +95,48 @@ func prefixAndSuffix(pattern string) (prefix, suffix string, err error) {
...
+		}
+		if IsExist(err) {
+			if try++; try < 10000 {
+				continue
+			}
+			return "", &PathError{Op: "mkdirtemp", Path: dir + string(PathSeparator) + prefix + "*" + suffix, Err: ErrExist}
+		}
+		if IsNotExist(err) {
+			_, serr := Stat(dir)
+			ine := IsNotExist(err)
```
Comment (dgryski) [inline src/os/tempfile.go]: Should this be checking `serr` instead of `err` ?

**Review by dgryski — CHANGES_REQUESTED** [review]

File: src/os/tempfile.go (hunk 1 above)
Comment (dkegel-fastly) [reply src/os/tempfile.go]: thanks.  That is partially reverted (and evidently broken) debugging code.  Fixing...

**Review by dgryski — APPROVED** [review]
LGTM

File: src/os/tempfile.go (hunk 1 above)
Comment (dkegel-fastly) [reply src/os/tempfile.go]: Squashed this fix in:

```
 		if IsNotExist(err) {
-			_, serr := Stat(dir)
-			ine := IsNotExist(err)
-			if ine {
-				return "", serr
+			if _, err := Stat(dir); IsNotExist(err) {
+				return "", err
 			}
 		}
```

to make it mirror the upstream test as intended.

**Comment by deadprogram** [comment]
Thank you @dkegel-fastly you are on a serious roll! Thanks for keeping up the review pace too @dgryski 

Now merging.

**Pull request merged**